src/devices/cpu/ht1130/ht1130.cpp

```diff
@@ -0,0 +1,1051 @@
+		return;
+	}
+
+	case 0b00101111: // RETI : Return from interrupt subroutine
+	{
+		m_pc = m_stackaddr;
+		if (m_stackcarry)
+			setcarry();
+		else
+			clearcarry();
+		return;
+	}
+
+	///////////////////////////////////////////////////////////////////////////////////////
+	// Timer Ops
+	///////////////////////////////////////////////////////////////////////////////////////
+
+	case 0b00111011: // MOV A,TMRH : Move timer high nibble to accumulator
+	{
+		const u8 data = gettimer_upper();
+		setacc(data);
+		return;
+	}
+
+	case 0b00111010: // MOV A,TMRL : Move timer low nibble to accumulator
+	{
+		const u8 data = gettimer_lower();
+		setacc(data);
+		return;
+	}
+
+	case 0b00111101: // MOV TMRH,A : Move accumulator to timer high nibble
+	{
+		const u8 acc = getacc();
+		settimer_upper(acc);
+		return;
+	}
+
+	case 0b00111100: // MOV TMRL,A : Move accumulator to timer low nibble
+	{
+		const u8 acc = getacc();
+		settimer_lower(acc);
+		return;
+	}
+
+	case 0b01000111: // (with 8-bit immediate) : TIMER XXH : Set immediate data to timer counter
+	{
+		const u8 oprand = fetch();
+		settimer(oprand);
+		return;
+	}
+
+	///////////////////////////////////////////////////////////////////////////////////////
+	// SOUND Ops (unimplemented)
+	///////////////////////////////////////////////////////////////////////////////////////
+
+	case 0b01001011: // SOUND A : Activate SOUND channel with accumulator
+	{
+		LOGMASKED(LOG_UNHANDLED_SOUND_OPS, "SOUND A");
+		return;
+	}
+
+	case 0b01001001: // SOUND LOOP : Turn on sound repeat cycle
+	{
+		LOGMASKED(LOG_UNHANDLED_SOUND_OPS, "SOUND LOOP");
+		return;
+	}
+
+	case 0b01001010: // SOUND OFF : Turn off sound
+	{
+		LOGMASKED(LOG_UNHANDLED_SOUND_OPS, "SOUND OFF");
+		return;
+	}
+
+	case 0b01001000: // SOUND ONE : Turn on sound 1 cycle
+	{
+		LOGMASKED(LOG_UNHANDLED_SOUND_OPS, "SOUND ONE");
+		return;
+	}
+
+	case 0b01000101: // (with 4 bit immediate) : SOUND n : Activate SOUND channel n
+	{
+		u8 oprand = fetch() & 0x0f;
+		LOGMASKED(LOG_UNHANDLED_SOUND_OPS, "SOUND %d", oprand);
+		return;
+	}
+
+	///////////////////////////////////////////////////////////////////////////////////////
+	// other Ops
+	///////////////////////////////////////////////////////////////////////////////////////
+
+	case 0b00110111: // (with 00111110) : HALT : Halt system clock
+	{
+		const u8 oprand = fetch();
+		if (oprand == 0b00111110) // this is a 'NOP' must HALT always be followed by NOP to work?
+		{
+			m_inhalt = 1;
+		}
+		else
+		{
+			LOGMASKED(LOG_UNHANDLED_OPS, "<ill HALT %02x %02x>", inst, oprand);
+		}
+		return;
+	}
+
+	default:
+	{
+		LOGMASKED(LOG_UNHANDLED_OPS, "<ill %02x>", inst);
+		return;
+	}
+	}
+}
+
+void ht1130_device::execute_run()
+{
+	while (m_icount > 0)
+	{
+		debugger_instruction_hook(m_pc);
+
+		if (!m_inhalt)
+		{
+			do_op();
+		}
+		else
+		{
+			m_icount--;
+		}
+
+		if (m_timer & 0x100)
+		{
+			m_timer -= 0x100;
+			m_timerover = 1; // can also generate an interrupt
+		}
```

Comment:
Is this going to work properly with multi-cycle instructions?  As far as I can tell, the timer counts clock cycles, not instructions.

Reply:
My understanding of the datasheet is that any 2 byte instruction takes 2 cycles, and the timer is cycle based.  we increase timer when a byte is fetched, if it has overflowed by the time it reached here we set flag.

I believe it is correct, assuming my understanding is correct.

Reply:
Oh, this is the timer overflow check, not the timer tick.